### DISCUSSION TITLE
What is the correct TypeScript type for a q-file element?

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12240] [General - Components / Directives / etc] [answered]
I have a `q-file` with a template ref, it is setup like this:

```
<template>
    <q-file
      ref="myFileTemplateRef"
      v-model="file"
      @update:model-value="fileUpdate"
    />
</template>

<script setup lang="ts">
import { ref, Ref } from 'vue';
const myFileTemplateRef: Ref<HTMLInputElement | undefined> = ref(undefined);
myFileTemplateRef.value.pickFiles(); //<--- ERROR HERE
</script>
```

But I get a TypeScript error saying `Property 'pickFiles' does not exist on type 'HTMLInputElement'.ts(2339)`.

So I assume `HTMLInputElement` is the wrong type.

What is the correct type for `q-file`?

If there is no type, how do I create one?

**Comment by BenJackGill** [comment]
I'm not sure if this is correct, but I tried to extending the HTMLInputElement type as per below.

```
interface QFile extends HTMLInputElement {
    pickFiles: () => void;
}

const myFile: Ref<QFile | undefined> = ref(undefined);

if (myFile.value) {
    myFile.value.pickFiles();
}
```

The error has gone away, but I'm not sure if this is 100% correct.

The shape of the pickFiles method in the [q-file Quasar docs][1] has this `pickFiles ([evt]) => void 0`.

Which I think indicates it takes an event? Or maybe an event array?

Either way, I couldn't get that working so I just removed the argument from the type. Please chime in if you know the correct way of doing it.


  [1]: https://quasar.dev/vue-components/file-picker#introduction

**Comment by yusufkandemir** [comment, marked as answer]
The correct type for QFile is.... `QFile` 🙂

```html
<template>
  <q-file ref="fileRef" v-model="file" @update:model-value="fileUpdate" />

  <q-btn label="Click me" @click="onClick" />
</template>

<script lang="ts" setup>
import { ref, Ref } from 'vue';
import { QFile, Notify } from 'quasar';

// ...
// "as Ref<...>" is to avoid the need for checking if it's defined.
// We know it will always be defined after mounting, inside event listeners, etc.
// So, we take the responsibility to avoid some annoying and distracting checks.
const fileRef = ref() as Ref<QFile>;

function onClick (event: MouseEvent) {
  Notify.create({ message: 'You may now pick some files' });

  fileRef.value.pickFiles(event);
}

// ...
</script>
```

**Reply by BenJackGill** [reply]
Thanks, that does seem obvious now! And your other tips for `as Ref<QFile>` and using the pickFiles event were also very helpful.
